### PR TITLE
Accept float in PVSystem.get_irradiance

### LINKED ISSUE #1338: pvsystem.Array.get_irradiance raises an error if solar_zenith is passed as a float
**Bug Description**
`pvsystem.get_irradiance` documentation says solar_zenith can be passed as a float or a Series. However, if `dni_extra` is not defined and `solar_zenith` is passed as float, then `solar_zenith.index` is not defined and the call to `irradiance.get_extra_irradiance` with `solar_zenith.index` as an argument fails.

**To Reproduce**
`test_sys = pvsystem.PVSystem()`
`test_sys.get_irradiance(0.0, 0.0, 0.0, 0.0, 0.0)`

**Error message** 
`AttributeError: 'float' object has no attribute 'index'`

**Expected behavior**
Return a DataFrame containing the irradiance components, as specified in [the documentation](https://pvlib-python.readthedocs.io/en/stable/generated/pvlib.pvsystem.Array.get_irradiance.html).

**Possible solutions**
1.  Ask for datetime if `dni_extra` is not passed as argument
2. Remove float from supported types for `solar_zenith`

**Versions:**
 - ``pvlib.__version__``: 0.9.0
 - ``pandas.__version__``: 1.3.3
 - python: 3.9

### PR DESCRIPTION
**Pull request by cwhanse**
- [x] Closes #1338
 - [X] I am familiar with the [contributing guidelines](https://pvlib-python.readthedocs.io/en/latest/contributing.html)
 - [x] Tests added
 - ~~[ ] Updates entries in [`docs/sphinx/source/reference`](https://github.com/pvlib/pvlib-python/blob/main/docs/sphinx/source/reference) for API changes.~~
 - [x] Adds description and name entries in the appropriate "what's new" file in [`docs/sphinx/source/whatsnew`](https://github.com/pvlib/pvlib-python/tree/main/docs/sphinx/source/whatsnew) for all changes. Includes link to the GitHub Issue with `` :issue:`num` `` or this Pull Request with `` :pull:`num` ``. Includes contributor name and/or GitHub username (link with `` :ghuser:`user` ``).
 - [x] New code is fully documented. Includes [numpydoc](https://numpydoc.readthedocs.io/en/latest/format.html) compliant docstrings, examples, and comments where necessary.
 - [x] Pull request is nearly complete and ready for detailed review.
 - [x] Maintainer: Appropriate GitHub Labels (including `remote-data`) and Milestone are assigned to the Pull Request and linked Issue.

If `solar_zenith` is not a Series with a DatetimeIndex, then the solar constant is used.

### CODE REVIEW

**Comment by cwhanse** [comment]
Working on this bug fix uncovers a related bug: the `pvsystem.PVSystem.get_irradiance` docstring says that irradiance parameters can be tuple of float, but that doesn't work:

```
system = pvsystem.PVSystem(surface_tilt=32, surface_azimuth=135)
irrads = {'dni': (900., 0.), 'ghi': (600., 0.), 'dhi': (100., 0.)}
zenith = 55.366831
azimuth = 172.320038
irradiance = system.get_irradiance(zenith,
                                   azimuth,
                                   irrads['dni'],
                                   irrads['ghi'],
                                   irrads['dhi'])
```

raises `ValueError: Length mismatch for per-array parameter`

So tuple of float of length N only works with a PVSystem with N Arrays.

I would like to edit the docstring to remove the promise that a tuple of floats will work. If more than one irradiance value is to be processed, they can be supplied in a Series. A tuple of Series, one Series for each Array, should still be accepted.

**Comment by kandersolar** [comment]
> So tuple of float of length N only works with a PVSystem with N Arrays.

Isn't that the intended behavior?  Tuples for these inputs are supposed to index over Arrays.  That example:

```python
irrads = {'dni': (900., 0.), 'ghi': (600., 0.), 'dhi': (100., 0.)}
irradiance = system.get_irradiance(..., irrads['dni'], irrads['ghi'], irrads['dhi'])
```

is simulating a single timestamp for system with two Arrays, where for some reason it is daytime for one Array and night for the other.  The specific values are not realistic, but the use case is still valid (IMHO).  I wouldn't call it a bug.

**Comment by cwhanse** [comment]
It would be simpler to describe behavior if we could state "tuples are distributed over Arrays". But that's not what happens when the system has only one Array: the tuple is treated the same as a Series of timestamped values.

If we want to keep the case of "tuple of float" I can put that back in the docstring as:

```
dni: float, tuple of float, Series or tuple of Series
    Direct normal irradiance [W/m2]


Notes
------
        Each of ``dni``, ``ghi``, and ``dni`` may be passed as a float, tuple of
        float, Series or tuple of Series. If passed as a float or Series, these
        values are used for every Array. If passed as a tuple of float and there
        is more than one Array, the tuple is distributed over the Arrays so the
        tuple length must be the number of Arrays. If passed as a tuple of Series
        the tuple is distributed over the Arrays and the tuple length must be
        the number of Arrays.
```

**Comment by kandersolar** [comment]
> It would be simpler to describe behavior if we could state "tuples are distributed over Arrays". But that's not what happens when the system has only one Array: the tuple is treated the same as a Series of timestamped values.

Can you provide an example?  I'm not seeing that behavior:

```python
from pvlib.pvsystem import PVSystem
import pandas as pd

system = PVSystem(surface_tilt=32, surface_azimuth=135)

for converter in [pd.Series, tuple]:
    irradiance = system.get_irradiance(
        solar_zenith=converter([45, 45]),
        solar_azimuth=converter([180, 180]),
        dni=converter([900, 900]),
        ghi=converter([600, 600]),
        dhi=converter([100, 100])
    )
    print(irradiance)
```

**Comment by cwhanse** [comment]
> Can you provide an example? I'm not seeing that behavior.

Neither am I. I must have been imagining what was originally intended by including "tuple of float" in the docstring.

To be sure we're seeing the same thing, that code produces `ValueError: Length mismatch for per-array parameter` for `converter=tuple`. Correct?

And this is the case we want to accomodate?

**Comment by kandersolar** [comment]
> To be sure we're seeing the same thing, that code produces `ValueError: Length mismatch for per-array parameter` for `converter=tuple`. Correct?

Correct.

> And this is the case we want to accomodate?

On the contrary, I think pvlib already behaves correctly. Tuples should get distributed over Arrays, so tuples of anything (float or Series) should only work when the length of the tuple equals the number of Arrays.  The case above shows that a single-Array PVSystem raises an error when given tuples of length 2.  That seems correct to me.  So what I am saying is that I think the "related bug" described in https://github.com/pvlib/pvlib-python/pull/2227#issuecomment-2378029879 is actually the intended behavior.

**Comment by cwhanse** [comment]
> Tuples should get distributed over Arrays, so tuples of anything (float or Series) should only work when the length of the tuple equals the number of Arrays. 

We are in agreement now. Thanks.

File: docs/sphinx/source/whatsnew/v0.11.2.rst [hunk 1]
```diff
@@ -20,7 +24,10 @@ Bug fixes
 Documentation
 ~~~~~~~~~~~~~
 * Edited docstrings for :py:func:`~pvlib.pvsystem.dc_ohms_from_percent` and
-  :py:func:`~pvlib.pvsystem.dc_ohmic_losses` for clarity. (:issue:`1601`, :pull:`2229`)
+  :py:func:`~pvlib.pvsystem.dc_ohmic_loss` for clarity. (:issue:`1601`, :pull:`2229`)
```
Comment (kandersolar) [inline docs/sphinx/source/whatsnew/v0.11.2.rst:27]: This should be `losses`, not `loss`.  Somehow this change keeps sneaking into PRs, I am not sure why.

File: docs/sphinx/source/whatsnew/v0.11.2.rst [hunk 2]
```diff
@@ -20,7 +24,10 @@ Bug fixes
...
+* Clarified that :py:meth:`~pvlib.pvsystem.PVSystem.get_irradiance` and
+  :py:meth:`~pvlib.pvsystem.Array.get_irradiance` do not accept tuple of float
+  as input. (:pull:`2227`)
```
Comment (kandersolar) [inline docs/sphinx/source/whatsnew/v0.11.2.rst:30]: Per the above discussion, I think this statement is not currently true, nor should it be a goal.  Tuple of float is the form corresponding to multiple arrays and a single timestamp of array-specific operating conditions:

```python
In [15]: arrays = [
    ...:     pvlib.pvsystem.Array(pvlib.pvsystem.FixedMount(surface_tilt=10)),
    ...:     pvlib.pvsystem.Array(pvlib.pvsystem.FixedMount(surface_tilt=20)),
    ...: ]
    ...: system = pvlib.pvsystem.PVSystem(arrays=arrays)
    ...: 
    ...: dni = (1000, 200)  # sunny on array 1, cloudy on array 2
    ...: dhi = (200, 300)
    ...: ghi = (800, 400)
    ...: zenith = 50
    ...: azimuth = 170
    ...: 
    ...: system.get_irradiance(zenith, azimuth, dni, ghi, dhi)
Out[15]: 
(OrderedDict([('poa_global', 992.7295708775583),
              ('poa_direct', 764.0235366741815),
              ('poa_diffuse', 228.7060342033768),
              ('poa_sky_diffuse', 227.1868095045976),
              ('poa_ground_diffuse', 1.519224698779198)]),
 OrderedDict([('poa_global', 482.67340206708127),
              ('poa_direct', 172.40899902290855),
              ('poa_diffuse', 310.26440304417275),
              ('poa_sky_diffuse', 307.2490340834682),
              ('poa_ground_diffuse', 3.0153689607045786)]))

```

File: pvlib/pvsystem.py [hunk 3]
```diff
@@ -323,13 +323,13 @@ def get_irradiance(self, solar_zenith, solar_azimuth, dni, ghi, dhi,
...
+        dni : float, Series, or tuple of Series
             Direct Normal Irradiance. [W/m2]
-        ghi : float or Series or tuple of float or Series
+        ghi : float, Series, or tuple of Series
             Global horizontal irradiance. [W/m2]
-        dhi : float or Series or tuple of float or Series
+        dhi : float, Series, or tuple of Series
             Diffuse horizontal irradiance. [W/m2]
-        dni_extra : float, Series or tuple of float or Series, optional
+        dni_extra : float, Series or tuple of Series, optional
```
Comment (kandersolar) [inline pvlib/pvsystem.py:332]: I think we should keep `tuple of float` as an option here

File: docs/sphinx/source/whatsnew/v0.11.2.rst [hunk 4]
```diff
@@ -21,6 +25,9 @@ Documentation
 ~~~~~~~~~~~~~
 * Edited docstrings for :py:func:`~pvlib.pvsystem.dc_ohms_from_percent` and
   :py:func:`~pvlib.pvsystem.dc_ohmic_losses` for clarity. (:issue:`1601`, :pull:`2229`)
+* Corrected a bug in :py:meth:`~pvlib.pvsystem.PVSystem.get_irradiance` and
+  :py:meth:`~pvlib.pvsystem.Array.get_irradiance` when `solar_zenith` is passed
+  float as input. (:issue:`1338`, :pull:`2227`)
```
Comment (cwhanse) [inline docs/sphinx/source/whatsnew/v0.11.2.rst:30]:
```suggestion
* Corrected a bug in :py:meth:`~pvlib.pvsystem.PVSystem.get_irradiance` and
  :py:meth:`~pvlib.pvsystem.Array.get_irradiance` when `solar_zenith` is passed
  float as input. (:issue:`1338`, :pull:`2227`)
```

**Comment by cwhanse** [comment]
@kandersolar I think I have addressed the review comments

File: pvlib/pvsystem.py [hunk 5]
```diff
@@ -1112,16 +1119,27 @@ def get_irradiance(self, solar_zenith, solar_azimuth, dni, ghi, dhi,
...
         if albedo is None:
             albedo = self.albedo
 
-        # not needed for all models, but this is easier
-        if dni_extra is None:
+        # dni_extra is not needed for all models, but this is easier
+        if (dni_extra is None and hasattr(solar_zenith, 'index')
+                and isinstance(solar_zenith.index, pd.DatetimeIndex)):
             dni_extra = irradiance.get_extra_radiation(solar_zenith.index)
+        else:
```
Comment (kandersolar) [inline pvlib/pvsystem.py:1140]:
```suggestion
        elif dni_extra is None:
```

File: pvlib/pvsystem.py [hunk 6]
```diff
@@ -1112,16 +1119,27 @@ def get_irradiance(self, solar_zenith, solar_azimuth, dni, ghi, dhi,
...
+            # use the solar constant
+            dni_extra = 1367
```
Comment (kandersolar) [inline pvlib/pvsystem.py:1142]:
```suggestion
        # dni_extra is not needed for all models, but this is easier
        if dni_extra is None:
            if (hasattr(solar_zenith, 'index') and
                isinstance(solar_zenith.index, pd.DatetimeIndex)):
                dni_extra = irradiance.get_extra_radiation(solar_zenith.index)
            else:
                # use the solar constant
                dni_extra = 1367
```

an alternative

**Review by kandersolar — APPROVED** [review]
I think a small logic tweak is needed, otherwise LGTM

**Comment by kandersolar** [comment]
Thanks @cwhanse

**Pull request merged**